ledger/src/blockstore.rs

```diff
@@ -1702,6 +1734,106 @@ impl Blockstore {
         })
     }
 
+    fn get_any_valid_slot_entries(&self, slot: Slot, start_index: u64) -> Vec<Entry> {
+        // lowest_cleanup_slot is the last slot that was not cleaned up by
+        // LedgerCleanupService
+        let lowest_cleanup_slot = self.lowest_cleanup_slot.read().unwrap();
+        if *lowest_cleanup_slot > slot {
+            return vec![];
+        }
+
+        let slot_meta_cf = self.db.column::<cf::SlotMeta>();
+        let slot_meta = slot_meta_cf.get(slot).unwrap_or_default();
+        if slot_meta.is_none() {
+            return vec![];
+        }
+
+        let slot_meta = slot_meta.unwrap();
+        // Find all the ranges for the completed data blocks
+        let completed_ranges = Self::get_completed_data_ranges(
+            start_index as u32,
+            &slot_meta.completed_data_indexes[..],
+            slot_meta.consumed as u32,
+        );
+        if completed_ranges.is_empty() {
+            return vec![];
+        }
+
+        let entries: Vec<Vec<Entry>> = PAR_THREAD_POOL.with(|thread_pool| {
+            thread_pool.borrow().install(|| {
+                completed_ranges
+                    .par_iter()
+                    .map(|(start_index, end_index)| {
```

Comment:
Each of these `end_index` should be a DATA_COMPLETE because `get_completed_data_ranges` should only return complete ranges. 

I think here instead of calling `get_any_valid_entries_in_data_block`, you an just call `get_entries_in_data_block`, and let `entries ` be of type `Vec<Result<Vec<Entry>>`. Then iterate over and take all good results up until the first error (since ReplayStage wouldn't have played anything beyond the first error anyways)